Add NAND and NOR number operations

Kryonn opened the pull request:
I was taking a look at logical operations and noticed that there is no function for NAND and NOR operations. This type of operation is most commonly used in projects related to digital systems, but since we have the logic category, I thought it would be a good idea to add these two to make it more complete. For now, I've only done this for number variables, but the idea is to expand the types of variables with subsequent PRs.

josdejong commented:
Thanks @Kryonn! It makes sense to add two two variations too. 

> For now, I've only done this for number variables, but the idea is to expand the types of variables with subsequent PRs.

I indeed think though that we should not only support them for the plain numbers, but also for mathjs in general: add these two new functions to [`/src/function/logical`](https://github.com/josdejong/mathjs/tree/develop/src/function/logical).

Kryonn commented:
> Just to be sure: when implementing functions, there are a couple of extra things that need to be taken care of. This is documented here and may be of help: https://github.com/josdejong/mathjs/blob/develop/README.md#implementing-a-new-function

I decided to start with the NOR function, and then do the NAND function. So, for the NOR function, I checked and applied the steps mentioned. However, when I went to test the function, I got two errors that I can't seem to resolve. The first error was related to the return of the toTex function. I didn't quite understand how we define the return for a given function. The second problem was related to the operation between two sparse matrices. I couldn't quite understand the error message. Could you help me?

###### error message
```
26 passing (27ms)
  2 failing

  1) nor
       should LaTeX nor:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
actual expected

'\\mathrm{nor}\\left(1,\\2\\right)'

      + expected - actual

      -\mathrm{nor}\left(1,2\right)
      +\left(1\2\right)

      at Context.<anonymous> (file:///C:/Users/Kryoj/OneDrive/Documents/GitHub/mathjs/test/unit-tests/function/logical/nor.test.js:234:12)
      at process.processImmediate (node:internal/timers:485:21)

  2) nor
       SparseMatrix
         should nor sparse matrix - sparse matrix:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly deep-equal:
+ actual - expected

  SparseMatrix {
    _datatype: undefined,
+   _index: [],
-   _index: [
-     0
-   ],
    _ptr: [
      0,
+     0,
+     0
-     1,
-     1
    ],
    _size: [
      2,
      2
    ],
+   _values: []
-   _values: [
-     true
-   ]
  }

      + expected - actual

       {
         "_datatype": [undefined]
      -  "_index": []
      +  "_index": [
      +    0
      +  ]
         "_ptr": [
           0
      -    0
      -    0
      +    1
      +    1
         ]
         "_size": [
           2
           2
         ]
      -  "_values": []
      +  "_values": [
      +    true
      +  ]
       }
```

Kryonn commented:
I finished the NOR function, it’s now working correctly and without any documentation issues.

As for the NAND function, I tried to implement it, but I wasn’t able to fix the test errors. From what I can tell, it would be necessary to create new matrix function application algorithms (matAlgoXXXX), which seems a bit complicated.

So, I’ll wrap up my contribution here. I hope I was able to help, and thank you for your support!

gwhitney commented:
I have no objection to adding `nand` and `nor` functions to mathjs, but is there any motivation for writing new internal code as opposed to just implementing them internally as `not(and(x, y))` and `not(or(x, y))`? is it the inefficiency of passing over array arguments twice, or something else? If the former is the only real concern, then I'd recommend implementing only the matrix algorithms directly, with or without the matAlgo functions as appropriate, and forwarding all other arguments to `not(and(x,y))` or `not(or(x,y))`. In particular, since `x nor y === (not x) and (not y)` the "nor" matrix by scalar operations should short-circuit when the scalar is true like the "and" ones do when the scalar is false. (And similarly, the matrix implementation for "or" is actually your best template when implementing `L nand M === (not L) or (not M)`.

Finally, my writing those expressions above makes me wonder whether `nand` and `nor` should be added as named infix operators like `or` and `and`? Again, I don't see any reason why they should be treated differently from the other logical operations, so I would recommend doing so.

@Kryonn, looking forward to your feedback on these points, and whether based on them you want to restart your efforts to get these operations into mathjs? If not, please be aware that due to limited time on the part of the core mathjs team, this PR would likely be converted to a discussion on adding "nand" and "nor" to mathjs, with a pointer to the closed PR as a starting place for anyone who is motivated to take up the effort to add these operations. It's just unfortunate that this PR as it stands isn't quite ready for merging (and in particular I have a tough time imagining merging `nor` without `nand`).

Kryonn commented:
> I have no objection to adding `nand` and `nor` functions to mathjs, but is there any motivation for writing new internal code as opposed to just implementing them internally as `not(and(x, y))` and `not(or(x, y))`? is it the inefficiency of passing over array arguments twice, or something else? If the former is the only real concern, then I'd recommend implementing only the matrix algorithms directly, with or without the matAlgo functions as appropriate, and forwarding all other arguments to `not(and(x,y))` or `not(or(x,y))`. In particular, since `x nor y === (not x) and (not y)` the "nor" matrix by scalar operations should short-circuit when the scalar is true like the "and" ones do when the scalar is false. (And similarly, the matrix implementation for "or" is actually your best template when implementing `L nand M === (not L) or (not M)`.
> 
> Finally, my writing those expressions above makes me wonder whether `nand` and `nor` should be added as named infix operators like `or` and `and`? Again, I don't see any reason why they should be treated differently from the other logical operations, so I would recommend doing so.
> 
> @Kryonn, looking forward to your feedback on these points, and whether based on them you want to restart your efforts to get these operations into mathjs? If not, please be aware that due to limited time on the part of the core mathjs team, this PR would likely be converted to a discussion on adding "nand" and "nor" to mathjs, with a pointer to the closed PR as a starting place for anyone who is motivated to take up the effort to add these operations. It's just unfortunate that this PR as it stands isn't quite ready for merging (and in particular I have a tough time imagining merging `nor` without `nand`).

I actually thought about doing it that way, but it seemed a bit too easy to be true. Anyway, I rewrote the NOR function and implemented the NAND function, both built by applying the OR and AND functions followed by the NOT function.

Kryonn reviewed (commented):
I finished the tasks. \o/

gwhitney commented on src/expression/parse.js:
```diff
@@ -788,15 +792,17 @@ export const createParse = /* #__PURE__ */ factory(name, dependencies, ({
   /**
    * logical and, 'x and y'
+   * logical nand, 'x nand y'
    * @return {Node} node
```
As I mentioned in earlier review, `nand` is actually a form of `or` (since `A nand B == (not A) or (not B)`). So `nand` should go in the logical or precedence group (maybe rename it parseLogicalOrNand?? that's optional). And conversely `A nor B == (not A) and (not B)` is a form of `and` so should go in this precedence group, which you could optionally rename `parseLogicalAndNor`. Thanks!

gwhitney commented on test/unit-tests/expression/parse.test.js:
```diff
@@ -2328,6 +2369,13 @@ describe('parse', function () {
         assert.strictEqual(parseAndEval('(true or true) and false'), false)
       })
 
+      it('should respect precedence between logical nand and or', function () {
```
Nice range of tests! This precedence test, though, needs adjusting because the cases you happened to choose don't distinguish the precedence options. For example, both of these expressions are true:  `(false nand true) or true` and `false nand (true or true)`. For testing precedence, you need an expression for which the result depends on how it is parenthesized. For example, for a nand - or combo like the first one, you could use `true nand false or true`, since `(true nand false) or true` is true, but `true nand (false or true)` is false.  You should find operands that show the same contrast for the other combos you test. Overall, those combos should include at least nand - and, nand-or, nor - and, and nor - or.  Thanks!

gwhitney commented:
That's fine, just looking for clarity. If you're able to handle the comments on what you've done so far, we should be able to get this PR merged before long.

josdejong commented:
@Kryonn did you see the last comment from Glen? Is it possible for you to have a last look into the tests?
